presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java

```diff
@@ -152,45 +223,57 @@ public static OGCGeometry deserialize(Slice shape)
         }
         BasicSliceInput input = shape.getInput();
 
-        int spatialReferenceId = input.readInt();
-        SpatialReference spatialReference = null;
-        if (spatialReferenceId != SPATIAL_REFERENCE_UNKNOWN) {
-            spatialReference = SpatialReference.create(spatialReferenceId);
-        }
-
         // GeometryCollection: spatialReferenceId|len-of-shape1|bytes-of-shape1|len-of-shape2|bytes-of-shape2...
         List<OGCGeometry> geometries = new ArrayList<>();
-        while (input.available() > 0) {
-            int length = input.readInt();
-            ByteBuffer buffer = input.readSlice(length).toByteBuffer().slice().order(LITTLE_ENDIAN);
-            Geometry esriGeometry = local().execute(0, Unknown, buffer);
-            OGCGeometry geometry = createFromEsriGeometry(esriGeometry, spatialReference);
-            geometries.add(geometry);
+
+        if (input.available() > 0) {
+            byte code = input.readByte();
+            while (input.available() > 0) {
+                OGCGeometry geometry = extractGeometry(code, input);
```

Comment:
`extractGeometry` converts `code` into `GeometryTypeName` and checks if it is `GEOMETRY_COLLECTION` on each loop iteration unnecessarily; let's check this once

```
if (input.readByte() != GEOMETRY_COLLECTION.ordinal()) {
    ByteBuffer buffer = input.slice().toByteBuffer().slice().order(LITTLE_ENDIAN);
    Geometry esriGeometry = local().execute(0, Unknown, buffer);
    return createFromEsriGeometry(esriGeometry, null);
}
else {
    <the original while loop>
}
```

Reply:
I put the check outside the loop. But I think
```
ByteBuffer buffer = input.slice().toByteBuffer().slice().order(LITTLE_ENDIAN);
Geometry esriGeometry = local().execute(0, Unknown, buffer);
return createFromEsriGeometry(esriGeometry, null);
```
appear in more than one places. Probably it's better to extract it to a method and reuse it.